### PR TITLE
chore: Add postgres dependency for server to startup

### PR DESCRIPTION
**Pull request by abhvsn**
## Description
As in the past we have seen the corruption of postgres DB which is being used for temporal we want to make sure we have a retry mechanism in place if:
1. `APPSMITH_DB_URL` is pointing to postgres url
2. Postgres is waiting in recovery mode

As per local testing when the docker container is abruptly shutdown via `docker rm -f {container_name}` or `docker kill {container_name}` or even via docker desktop we end up in state where postgres goes into recovery state. 
logs:
```
2024-09-27 08:02:49 backend stdout | SQL State  : 57P03
2024-09-27 08:02:49 backend stdout | Error Code : 0
2024-09-27 08:02:49 backend stdout | Message    : FATAL: the database system is starting up
```
Currently we have implemented polling mechanism, but we will keep looking for better alternative here if we can opt for.   

Note: 
1. With release dump this is taking ~300sec to get out of that state and start accepting the connections.
2. With the existing implementation without retries server dies down within 60sec.
```
INFO exited: backend (exit status 1; not expected)
INFO gave up: backend entered FATAL state, too many start retries too quickly
```
Reference doc: https://www.notion.so/appsmith/Postgres-critical-scenarios-668f49c96aef40308e24c2a8d6b1137c

/test Sanity

### :mag: Cypress test results

> [!TIP]
> 🟢 🟢 🟢 All cypress tests have passed! 🎉 🎉 🎉
> Workflow run: <https://github.com/appsmithorg/appsmith/actions/runs/11100944184>
> Commit: 9dbbe4b22ba12aa82c385ad0eef3cc3d4876f217
> <a href="https://internal.appsmith.com/app/cypress-dashboard/rundetails-65890b3c81d7400d08fa9ee5?branch=master&workflowId=11100944184&attempt=1" target="_blank">Cypress dashboard</a>.
> Tags: `@tag.Sanity`
> Spec:
> <hr>Mon, 30 Sep 2024 07:26:41 UTC

## Communication
Should the DevRel and Marketing teams inform users about this change?
- [ ] Yes
- [x] No

## Summary by CodeRabbit

- **New Features**
	- Introduced new functions for enhanced handling of PostgreSQL database connections, including availability checks and parameter extraction.
	- Added a new utility script for managing PostgreSQL connections.
- **Bug Fixes**
	- Implemented a retry mechanism for PostgreSQL server availability checks to ensure more reliable connections.
- **Tests**
	- Added unit tests to validate the functionality of the PostgreSQL parameter extraction logic.

### CODE REVIEW

**Review by pratapaprasanna — APPROVED** [review]

File: deploy/docker/fs/opt/appsmith/pg-utils.sh [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+    const parsedUrl = new url.URL(pgUri);
+
+    // Extract the pathname and remove the leading '/'
+    const db = parsedUrl.pathname.substring(1);
+
+    // Default the port to 5432 if it's empty
+    const port = parsedUrl.port || '5432';
+
+    console.log(\`\${parsedUrl.username || '-'} \${parsedUrl.password || '-'} \${parsedUrl.hostname} \${port} \${db}\`);
+  " "$conn_string")
```
[the remark this answers is not in the document]
Comment (mohanarpit) [reply deploy/docker/fs/opt/appsmith/pg-utils.sh:80]: @abhvsn Can you please modify the code to resolve this comment?

[the remark this answers is not in the document]
Comment (abhvsn) [reply deploy/docker/fs/opt/appsmith/pg-utils.sh:80]: @mohanarpit I am not able to understand the advantage for this. I have written the tests in my local to see the effect of this change but couldn't find anything, tests are written for following `APPSMITH_DB_URL` input for which the response was exactly same with or without quotes:
1. postgresql://user:password@localhost:5432/dbname
2. postgresql://user:p a s s w o r d@localhost:5432/dbname
3. postgresql://user:password@localhost:5432/db name

[the remark this answers is not in the document]
Comment (abhvsn) [reply deploy/docker/fs/opt/appsmith/pg-utils.sh:80]: https://github.com/coderabbitai can you please check the above comment, if I'm missing something?

File: deploy/docker/tests/test_pg-utils.sh [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+  echo "PG_DB_PORT: $PG_DB_PORT"
+  echo "PG_DB_NAME: $PG_DB_NAME"
+}
+
+# Run tests
+test_extract_postgres_db_params_valid_db_string
+test_extract_postgres_db_params_empty_dbname
+test_extract_postgres_db_params_with_spaces
+
+echo "All Tests Pass!"
```
[the remark this answers is not in the document]
Comment (abhvsn) [reply deploy/docker/tests/test_pg-utils.sh:61]: Let me do that in a separate PR can you raise an issue for this though?

**Review by AnaghHegde — APPROVED** [review]

**Pull request merged**